generic-ec-zkp/src/dlog_eq.rs

```diff
@@ -0,0 +1,277 @@
+//! # fn send<T>(_: T) {}
+//!
+//! // `x` is a secret discrete logarithm only known to prover
+//! let x = SecretScalar::<E>::random(&mut OsRng);
+//!
+//! // `X` is a public point corresponding to `x`
+//! let X = Point::generator() * &x;
+//! // `H` is some point known to both parties
+//! let H = Point::generator() * Scalar::random(&mut OsRng);
+//! // `Z` is the object of the proof
+//! let Z = H * &x;
+//!
+//! let data = interactive::Data {
+//!     pub_share: X,
+//!     base: H,
+//!     exp: Z,
+//! };
+//!
+//! // Prover chooses a nonce for itself
+//! let r = Scalar::random(&mut OsRng);
+//! // Prover commits to the data
+//! let commitment = interactive::commit_data(data, r);
+//! // Prover sends this commitment to the verifier
+//! send(commitment);
+//!
+//! // Verifier sends a challenge to the prover
+//! let challenge = Scalar::random(&mut OsRng);
+//! send(challenge);
+//!
+//! // Prover receives the challenge and computes the proof of equality `H * x = Z`
+//! let proof = interactive::prove(r, challenge, &x);
+//! // This proof is sent to the verifier
+//! send(proof);
+//!
+//! // Verifier checks that the proof is correct
+//! interactive::verify(data, commitment, challenge, proof)
+//!     .expect("Verification failed!");
+//! ```
+
+use generic_ec::{Curve, Point};
+
+/// Object of the proof
+#[derive(Debug, Clone, Copy)]
+#[cfg_attr(feature = "udigest", derive(udigest::Digestable), udigest(bound = ""))]
+pub struct Data<E: Curve> {
+    /// `X`, the value inside `log_G`
+    pub pub_share: Point<E>,
+    /// `H`, the base for the second logarithm
+    pub base: Point<E>,
+    /// `Z`, the value inside the second logarithm `log_H`
+    pub exp: Point<E>,
+}
+
+/// Functions for interactive protocol for the proof. See usage example in
+/// [`dlog_eq`](crate::dlog_eq)
+///
+/// Based on "Zero-Knowledge Proofs Notes", 2024 by Jorge L. Villar
+pub mod interactive {
+    use generic_ec::{Curve, Point, Scalar, SecretScalar};
+
+    /// Commitment to `H`, sent in the first round
+    pub type Commitment<E> = (Point<E>, Point<E>);
+    /// Challenge, sent to the Prover in the first round
+    pub type Challenge<E> = Scalar<E>;
+    /// Proof data, sent by prover to the verifier
+    pub type Proof<E> = Scalar<E>;
+    pub use super::Data;
+    pub use super::InvalidProof;
+
+    /// First round of the protocol: commit to `H` by itself. Produces
+    /// [`Commitment`] to be sent
+    ///
+    /// - `r` - random nonce. Can be produced by `Scalar::random(&mut rng)`
+    pub fn commit<E: Curve>(base: Point<E>, r: Scalar<E>) -> Commitment<E> {
+        (Point::generator() * r, base * r)
+    }
+    /// First round of the protocol: commit to well-constructed [`Data`] by
+    /// itself. Produces [`Commitment`] to be sent
+    ///
+    /// - `r` - random nonce. Can be produced by `Scalar::random(&mut rng)`
+    pub fn commit_data<E: Curve>(data: Data<E>, r: Scalar<E>) -> Commitment<E> {
+        commit(data.base, r)
+    }
```

Comment:
Why did you choose a new way of defining commit function? In all other proofs, commit function takes public data and source of randomness and produces the commitment

Reply:
1. Because I was mirroring how it's done in the tBLS paper, where `r` is generated externally.
2. Because otherwise we would have to make an explicit private commitment structure. Actually this might be a good thing
3. I felt like having non_interactive::commit be independent of rng dependency is a nice goal

I don't have a strong opinion on any of this though, just giving my thought process

Reply:
I think, ideally we should make the API hard to misuse, in particular, exposed functions should contain most of the workflow. If we expect caller to do something correctly and provide it as inputs, that's where something can go wrong, so it's easier to misuse

Reply:
Yeah you're actually right, the only thing you can do with those nonces is generate and use them in a proof once, so they should be rng. Fixed